src/jaxsim/rbda/contacts/rigid.py

```diff
@@ -0,0 +1,509 @@
+            model: The `JaxSimModel` instance.
+            data: The `JaxSimModelData` instance.
+            link_external_forces: Optional `(n_links, 6)` matrix of external forces acting on the links,
+                expressed in the same representation of data.
+
+        Returns:
+            A tuple containing the contact forces.
+        """
+
+        # Import qpax just in this method
+        import qpax
+
+        link_external_forces = (
+            link_external_forces
+            if link_external_forces is not None
+            else jnp.zeros((model.number_of_links(), 6))
+        )
+
+        # Compute kin-dyn quantities used in the contact model
+        with data.switch_velocity_representation(VelRepr.Mixed):
+            M = js.model.free_floating_mass_matrix(model=model, data=data)
+            J_WC = js.contact.jacobian(model=model, data=data)
+            W_H_C = js.contact.transforms(model=model, data=data)
+        terrain_height = jax.vmap(self.terrain.height)(position[:, 0], position[:, 1])
+        terrain_normal = jax.vmap(self.terrain.normal)(position[:, 0], position[:, 1])
+        n_collidable_points = model.kin_dyn_parameters.contact_parameters.point.shape[0]
+
+        # Compute the activation state of the collidable points
+        inactive_collidable_points, (delta, delta_dot) = RigidContacts.detect_contacts(
+            W_o_C=position,
+            W_o_dot_C=velocity,
+            terrain_height=terrain_height,
+            terrain_normal=terrain_normal,
+        )
+
+        delassus_matrix = RigidContacts._delassus_matrix(M=M, J_WC=J_WC)
+        # Make it symmetric if not
+        delassus_matrix = jax.lax.cond(
+            jnp.allclose(delassus_matrix, delassus_matrix.T),
+            lambda G: G,
+            lambda G: (G + G.T) / 2,
+            delassus_matrix,
+        )
+        # Add regularization for better numerical conditioning
+        delassus_matrix = delassus_matrix + 1e-6 * jnp.eye(delassus_matrix.shape[0])
+
+        references = js.references.JaxSimModelReferences.build(
+            model=model,
+            data=data,
+            velocity_representation=data.velocity_representation,
+            link_forces=link_external_forces,
+        )
+
+        with references.switch_velocity_representation(VelRepr.Mixed):
+            nu_dot_free_mixed = RigidContacts._compute_mixed_nu_dot_free(
+                model, data, references=references
+            )
+
+        free_contact_acc = RigidContacts._linear_acceleration_of_collidable_points(
+            model,
+            data,
+            nu_dot_free_mixed,
+        ).flatten()
+        # Compute stabilization term
+        baumgarte_term = RigidContacts._compute_baumgarte_stabilization_term(
+            inactive_collidable_points=inactive_collidable_points,
+            delta=delta,
+            delta_dot=delta_dot,
+            K=self.parameters.K,
+            D=self.parameters.D,
+        ).flatten()
```

Comment:
In your formulation, you are working in mixed representation. This means that also the contact forces are ${}_{C[W]} \boldsymbol{f}$. It seems to me that this stabilization term only injects forces along the z direction, but they are computed using $(\delta, \dot{\delta})$ that are projected along the terrain's normal.

I just want to double check if this sounds good to you. In this case, maybe it would be more consistent to use $\mathbf{h}$ (the penetration along $z$) instead of $\delta$. What do you think? Note that they differ only on non-flat terrains.

Reply:
Yes, I agree with you that this Buamgarte term in only injecting forces in the vertical direction, hence it is better to align the notation to use h instead of delta. 

I'm not sure that this is the most useful way to use this stabilization term, but I would postpone to future PRs the role of updating this formulation.